arch: arm64: initialize IRQ stack and canary

jharris-intel opened the pull request:
As per https://github.com/zephyrproject-rtos/zephyr/pull/30676#discussion_r574319395.

Fair warning: this is my first PR on this project. It's _entirely_ likely that I've missed things.

(Also: the [documentation](https://docs.zephyrproject.org/latest/contribute/index.html) doesn't seem to mention that `uncrustify` shouldn't be run on .S files.)

Signed-off-by: James Harris <james.harris@intel.com>

galak commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -63,6 +66,56 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 
 	ldr	x0, =(z_interrupt_stacks)
+
+#define IS_POWER_OF_TWO(X) ((X) && (((X) & ((X) - 1)) == 0))
```
Might be good to put `IS_POWER_OF_TWO` in include/sys/util.h, already has a C inline func version.

jharris-intel replied:
I could do that.

Some questions around how you do things "around here" that I haven't really seen answered elsewhere, likely somewhat silly ones:

1. If I did so, would I keep the title as arch: arm64 seeing as I modified shared code? Or?
2. Add `IS_POWER_OF_TWO` as a separate commit (i.e. rewrite history of this PR such that the first commit adds `IS_POWER_OF_TWO` and the second uses it)? Or as part of the same commit? Or as a different PR?
3. Does Zephyr care about tests passing on intermediate commits in a single PR? (Some projects do, some don't).

galak replied:
> * If I did so, would I keep the title as arch: arm64 seeing as I modified shared code? Or?
> 
For something trivial no need to change the title.

> * Add `IS_POWER_OF_TWO` as a separate commit (i.e. rewrite history of this PR such that the first commit adds `IS_POWER_OF_TWO` and the second uses it)? Or as part of the same commit? Or as a different PR?
> 

Again, for something trivial its up to you to decide.  It makes sense to keep in the same PR since it shows the code that is using it as well.

> * Does Zephyr care about tests passing on intermediate commits in a single PR? (Some projects do, some don't).

We do care about this, but aren't strictly enforcing it.

jharris-intel replied:
Thanks for the answers!

jharris-intel replied:
I may have overkilled this somewhat. Could you please take another look?

jharris-intel commented:
I'll need to fix this dev box's author name before this can be merged.

- [ ] Fix author name from `jharri` to `James Harris`

ioannisg commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -15,6 +15,9 @@
 #include "vector_table.h"
 #include "macro_priv.inc"
+#ifdef CONFIG_STACK_SENTINEL
+#include <kernel_structs.h> // For STACK_SENTINEL definition
```
hm, i think we 're not using '//' comment-style at all

jharris-intel replied:
My apologies - I'm used to a style that does.

- [ ] Change `//`-style comments to `/*`-style.

ioannisg commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -63,6 +66,56 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
+	 * to the next multiple of ARCH_STACK_PTR_ALIGN, but that would
+	 * just be wasting space.
+	 */
+	b .
```
I am not sure I like the spinning here - it gives little information to the developer. I understand, though, that we are too early in the boot sequence and proper cpu fault reporting mechanism is not established.

Also, it is not clear why this is under CONFIG_ASSERT; after all, there is no explicit assert statement here.

I'd like others to comment here (@carlocaione @andyross ) too; IMO, we should be using build time checks for ISR stack size/alignment (and any other thread stack definition).

Unless I am missing something, defining thread and isr stack sizes using the right kernel macros (and their arch-specific extensions) should do the work, so no need for run-time checks.

jharris-intel replied:
> I am not sure I like the spinning here - it gives little information to the developer.

I am in agreement here. It's mainly a spin here because I don't know of a better alternative:

1. C (and ASM) does not consider the address of a global a compile-time-known construct, so I cannot check this at compile time without resorting to non-standard constructs like `__alignof__`.
2. Checking this later in runtime does not accomplish anything, as by the time we would check we'd _already_ have jumped off into the weeds (likely a Stack Alignment fault)
3. At this point in bootflow an actual assert call will do nothing other than obscure the source of the failure.

(Also: is the expectation for Zephyr that someone will be debugging early boot without access to a debug probe of some sort? With a debug probe this gives fairly clear information to the developer so far as I can tell. We're sitting at a branch-to-self, find where in the source the address corresponds to, see the comment in the source immediately before it as to what this is checking and why...)

> it is not clear why this is under CONFIG_ASSERT; after all, there is no explicit assert statement here.

It's conceptually an assert, and would be an actual assert if I had any confidence at this point in bootflow that an assert would do anything other than further obscure the actual failure. Unfortunately...

Is there a better pattern I should use for this sort of early-boot check?

> Unless I am missing something, defining thread and isr stack sizes using the right kernel macros (and their arch-specific extensions) should do the work, so no need for run-time checks.

Fair enough. I tend to be fairly overzealous about runtime checks in asserts-enabled builds in places that aren't performance-sensitive. (i.e. ideally _both_ a compile-time and a runtime check, with the compile-time check always enabled and the runtime check enabled for `CONFIG_ASSERT`.).

This is mainly because I've had too many cases in the past where a compile-time check got inadvertently broken or removed (e.g. moved to a file that isn't compiled in a specific build) and I then spent a bunch of time trying to rewind time and figuring out what happened before things jumped off into the weeds.

jharris-intel committed:
toolchain: add Z_IS_POWER_OF_TWO

Single-evaluation version of IS_POWER_OF_TWO.

Signed-off-by: James Harris <james.harris@intel.com>

jharris-intel committed:
tests/unit/util: add Z_IS_POWER_OF_TWO tests

Add a basic smoketest to ensure Z_IS_POWER_OF_TWO works as expected.

Signed-off-by: James Harris <james.harris@intel.com>

jharris-intel committed:
sys: util: Add IS_POWER_OF_TWO

Zephyr has is_power_of_two, however this is a function and as such
not suitable for compile-time constructs. Add IS_POWER_OF_TWO and
rewrite is_power_of_two using it.

Signed-off-by: James Harris <james.harris@intel.com>

jharris-intel committed:
tests/unit/util: Add IS_POWER_OF_TWO tests

Add a few basic known-answer tests for IS_POWER_OF_TWO, covering
a few known or percieved edge cases (high bit set, zero, etc.)

Signed-off-by: James Harris <james.harris@intel.com>

jharris-intel committed:
arch: arm64: initialize IRQ stack and canary

As per https://github.com/zephyrproject-rtos/zephyr/pull/30676#discussion_r574319395.

Signed-off-by: James Harris <james.harris@intel.com>

jharris-intel commented:
@ioannisg / @carlocaione - ping?

I'm wondering how to resolve the spinning @ioannisg brought up above; this is also still waiting on reviewers in general.

carlocaione commented:
@jharris-intel I didn't forget about this but I was waiting for #32394 to be merged so that I can look at this again in the light of the new reset code.

jharris-intel commented:
Makes sense, thank you @carlocaione!

cfriedt commented:
@jharris-intel - please rebase and resolve conflicts. If review comments have been resolved, please mark them resolved.

@carlocaione / @ioannisg - please re-review.
